Implementation for mpio driver with selection I/O.

vchoi-hdfgroup opened the pull request:
This is a draft for preliminary review.

vchoi-hdfgroup committed:
Implementation for mpio driver with selection I/O.
This is a draft for preliminary review.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct CI errors.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

fortnern commented on src/H5FDint.c:
```diff
@@ -1555,7 +1562,7 @@ H5FD__write_selection_translate(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, ui
 
             /* Check if we're using vector I/O */
-            if (use_vector) {
+            if (!skip_vector_cb && use_vector) {
```
Should use skip_vector_cb to determine the value of use_vector on line 1460, then there will be no need to keep checking it

vchoi-hdfgroup replied:
Fixed

fortnern commented on src/H5FDint.c:
```diff
@@ -2392,6 +2434,431 @@ H5FD_sort_vector_io_req(hbool_t *vector_was_sorted, uint32_t _count, H5FD_mem_t
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5FD_selection_build_types(hbool_t io_op_write, uint32_t num_pieces, H5S_t **file_spaces, H5S_t **mem_spaces,
```
Having this function in a shared place is a little strange. I was originally thinking that the code in H5FDmpio.c and H5Dmpio.c shouldn't be shared. If it's working though it might be worth it. I'll continue thinking about a way to make this function more self-contained/generic/maintainable

fortnern replied:
After further thought, I think we should move this function to H5FDmpio.c, and return this functionality to being integrated into link_piece_collective_io (with the frees moved to before the I/O as suggested in my email). I think it'd also be a good idea to move the check for num_pieces out of the function (and into the read/write callbacks), and remove the size_i  and mpi_buds_base parameters (calculate them in the read/write_callbacks).

vchoi-hdfgroup replied:
I will move this function to H5FDmpio.c as indicated.
And I will restore link_piece_collective_io() back to its original state in H5Dmpio.c (with the frees moved to before the IO), is that correct?

fortnern replied:
That's correct, thanks!

vchoi-hdfgroup replied:
Changes made as indicated for H5FD_selection_build_types().
However, I still need to pass num_pieces to the routine because it is needed to malloc the arrays of mtype, ftype and displacement etc.  I can calculate mpi_bufs_base in the read/write callbacks, but I still need to pass the calculated info to H5FD_selection_build_types() for calculating displacements of the pieces.

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Fix use_vector on read/write_vector() != NULL and skip_vector_cb.
This is in response to one of the PR review comments.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
1) Restore link_piece_collective_io() back to its original state in H5Dmpio.c
(with the frees of the arrays moved to before the IO)
2) Move H5FD_selection_build_types() to H5FDmpio.c
This is in response to one of the PR review comments.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct errors made for previous checkin.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Correct CI errors from running autogeh.sh.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
This is the fix for a failure from one of the tests in testpar/t_2Gio.c when selection I/O is ON by default in the library.
a) Create internal functions H5FD_read/write_vector_from_selection() and H5FD_read/write_from_selection() that correspond to the 4 new APIs H5FDread/write_vector_from_selection() and H5FDread/write_from_selection().
--Call the internal functions H5FD_read/write_vector_from_selection in H5FD__mpio_read/write_selection() for independent xfer_mode.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
This is the fix for the test failure from running testpar/t_pread.c
when selection I/O is ON by default in the library.
--Undo the base address addition in internal routines before passing down to the mpio driver

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
This is the fix for a failure from dataset_writeAll() test in testpar/t_dsets.c (via testpar/testphdf5.c)
when selection I/O is ON by default in the library:
--need to check for fixed bufs when trying to figure out mpi_bufs_base in H5FD__selection_build_types().

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into mpio_develop

vchoi-hdfgroup committed:
Fix CI errors.

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Corrections to the previous fix and commit message:
This is the fix for a failure from test_link_chunk_io_sort_chunk_issue() test in
testpar/t_coll_md.c (via testpar/testphdf5.c) when selection I/O is ON by default
in the library:
--need to check for fixed bufs when trying to figure out mpi_bufs_base in H5FD__selection_build_types().

vchoi-hdfgroup committed:
Merge branch 'mpio_develop' of https://github.com/vchoi-hdfgroup/hdf5 into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Merge branch 'develop' into mpio_develop

github-actions[bot] committed:
Committing clang-format changes

mattjala committed:
Add note about HDF5_VOL_CONNECTOR to tools usage (#3159)

vchoi-hdfgroup committed:
Change HDassert, HDqsort, HDfprintf to assert, qsort, and fprintf accordingly.

github-actions[bot] committed:
Committing clang-format changes

derobins committed:
Rename HDsystem() to system() (#3197)

system() is only used in the iopipe test and the things it calls
(which are POSIX-y) are protected by an ifdef.

vchoi-hdfgroup committed:
Merge branch 'develop' into mpio_develop

vchoi-hdfgroup committed:
Change HDmalloc and HDfree to malloc and free.

vchoi-hdfgroup commented:
I am going to create a new PR based on the latest "develop" branch and incorporate the mpio driver changes to there.  I think it's easier that way instead of resolving all the conflicts/CI failures etc.  Then I will close this PR.

vchoi-hdfgroup commented:
See PR #3222
